src/main/java/org/folio/circulation/support/CirculationRulesClient.java

```diff
@@ -24,20 +23,17 @@
     root = context.getOkapiBasedUrl(policyPath);
   }
 
-  public void applyRules(
-    String loanTypeId,
-    String locationId,
-    String materialTypeId,
-    String patronGroup,
-    Handler<HttpClientResponse> responseHandler) {
+  public CompletableFuture<Result<Response>> applyRules(
+    String loanTypeId, String locationId, String materialTypeId,
+    String patronGroupId) {
 
     String circulationRulesQuery = queryParameters(loanTypeId, locationId,
-      materialTypeId, patronGroup);
+      materialTypeId, patronGroupId);
 
     log.info("Applying circulation rules for {}", circulationRulesQuery);
 
-    client.get(String.format("%s?%s", root, circulationRulesQuery),
-      responseHandler);
+    return client.toWebClient()
```

Comment:
I just think... Will it be easier to convert the `OkapiHttpClient` to interface and then provide different implementation of it? 

It is not so clear to me when `OkapiHttpClient` is being converted to VertxWebClient (which is actually another implementation).

Maybe I'm missing something.

Reply:
Excellent question.

As part of this switch over, most of the method signatures change significantly, from using a handler callback to return a future result. I did attempt that signature change in-situ and it was difficult to do incrementally.


I also didn't want to replace the okapi http client in situ, because it would mean mixing up methods that use the old vert.x HTTP client and the web client.

At the end of the 4 or 5 pull requests I have for this change over, the `OkapiHttpClient` changes from a class to an interface to allow for this to be changed more easily in the future, hopefully.

If you'd like a preview of where this switch over is heading, check out the https://github.com/folio-org/mod-circulation/tree/CIRC-308-use-web-client-for-put branch.

Does that explanation make sense?

Reply:
And yes, it would've been better if OkapiHttpClient had been an interface from the beginning. Although, it would still be difficult to replace incrementally.

Converting it to an interface, changing the interface to the new interface definition and then replacing the implementation would've been a better approach (and very Kent Beck esque). I rather got too far into the changes before that became apparent.

Reply:
Ok, thanks for explaining it.